src/stemmabench/algorithms/stemma_NJ.py

```diff
@@ -0,0 +1,161 @@
+            - distance (Callable, Optional): A function that takes 2 strings as parameters and returns a numeric value which is
+            the distance between the 2 strings.
+        """
+        super().__init__()
+        self._distance: Union[Callable, None] = distance
+        self._dist_matrix: Union[np.ndarray, None] = None
+
+    @property
+    def distance(self):
+        return self._distance
+
+    @property
+    def dist_matrix(self):
+        return self._dist_matrix
+
+    def compute(self,
+                folder_path: str,
+                distance: Union[Callable, None] = None) -> ManuscriptInTreeBase:
+        """Builds the stemma tree. If the distance is specified in function call it will surplant the existing distance if it exists.
+
+        ### Args:
+            - folder_path (str): The path to the folder containing the texts. The path specified here will surplant the previous path defined in constructor.
+            !!! All .txt files in this folder must be files containing Manuscript texts unless the file name contains the substring "edge" !!!
+            - distance (Callable, Optional): A function that takes 2 strings as parameters and returns a numeric value which is
+            the distance between the 2 strings.
+
+        Returns:
+            - Manuscript: The root of the stemma with the rest of its tree as its children.
+
+        Raises:
+            - RuntimeError: If the _distance attribute has not been specified in the constructor and has not been specified in the compute function call.
+        """
+        super().compute(folder_path)
+        if not self.distance and not distance:
+            raise RuntimeError(
+                "The _distance was not specified in the constructor, therefore the distance parameter must be specified.")
+        if distance:
+            self._distance = distance
+            self.dist(distance=distance)
+        else:
+            self.dist(distance=self._distance)
+        edges_dict, edges_list = self._build_edges()
+        out = ManuscriptInTreeEmpty(parent=None, recursive=Utils.dict_from_edge(
+            edge_list=edges_list), text_list=list(self.manuscripts.keys()))
+        out.set_edges(edges_dict)
+        return out
+
+    def dist(self, distance: Callable) -> None:
+        """Builds the distance matix based on the provided distance function and sets the attribute _dist_matrix.
+
+        ### Args:
+            - distance (Callable): A function that takes as parameters 2 strings and that returns the distance between them.
+
+        ### Raises:
+            - RuntimeError: If method is called and the manuscripts attribute is not set.
+        """
+        if not self.manuscripts:
+            raise RuntimeError(
+                "The manuscripts attribute needs the be set before the distance can be calculated.")
+        self._dist_matrix = np.ndarray(
+            (len(self.manuscripts), len(self.manuscripts)))
+        # TODO: use map instead
```

Comment:
tu peux peut être le faire maintenant ? 
avec scipy aussi tu peux gagner énormément en perf : https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance_matrix.html

Reply:
I'll do it in the next PR as I should be more familiar with parallel processing in python.